backend/src/crd/pkg/apis/viewer/v1alpha1/types.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Viewer is a specification for a Viewer resource.
+type Viewer struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec contains specifications that pertain to how the viewer is launched and
+	// managed by its controller.
+	Spec ViewerSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ViewerList is a list of Viewer resources.
+type ViewerList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of viewers.
+	Items []Viewer `json:"items"`
+}
+
+// ViewerType is the underlying type of the view. Currently, only Tensorboard is
+// explicitly supported by the Viewer CRD.
+type ViewerType string
+
+const (
+	// ViewerTypeTensorboard is the ViewerType constant used to indicate that the
+	// underlying type is Tensorboard. An instance named `instance123` will serve
+	// under /tensorboard/instance123.
+	ViewerTypeTensorboard ViewerType = "tensorboard"
+)
+
+// TensorboardSpec contains fields specific to launching a tensorboard instance.
+type TensorboardSpec struct {
+	// LogDir is the location of the log directory to be read by tensorboard, i.e.,
+	// ---log_dir.
+	LogDir string `json:"logDir"`
```

Comment:
This should be an array (slice?) of strings. Tensorboard supports comparing outputs by passing a list of logdirs, we're using this feature in run comparison.

Reply:
It's just a comma separated list right? Which can be passed in directly here. This may be more flexible, in that I'm not reinterpreting the arguments to Tensorboard. For example, TB allows naming some of those runs as well. Rather than structure that here, just pass in the expected format to Tensorboard. What do you think?

Reply:
That's fine I guess, although structuring it would give some value, because the structure is not very straightforward. Something like "name1:<encoded_path_1>,name2:<encoded_path_2>" is easy to mistype.

Reply:
True, but if we're planning on creating the viewer through UI, then we can try to construct it for the user there instead before sending the create request?

Reply:
Yup, we can have clients do it.